src/tree-set.satyg

```diff
@@ -136,7 +136,7 @@ end = struct
       let rv = r#value in
       let rl = r#left in
       let rr = r#right in
-      if height rr > height rl then
+      if height rr >= height rl then
```

Comment:
For the following tree T,

```
             9
          /     \
        4         11
      /   \      /  \
     2     6   10    13
    / \   / \        /
   1   3 5   8      12
  /         /
 0         7
```

`delete 13 T` used to result in the following tree, which does break the invariant.

```
             6
          /     \
        4          9
      /   \      /   \
     2     5     8    11
    / \         /    /  \
   1   3       7    10  12
  /
 0
```

I do not know how to test this in generic.saty, that's why I didn't update that file.

Reply:
One way to test this is to change the behavior of to-stream and to-list a bit to do bfs search instead of current, left prioritized (minimum-first maximum-last) search. I'm not sure which behavior is better for common use cases.

Reply:
I decided to change the behavior of to-stream and to-list. Now there is no guarantee in the order of their results.

Reply:
The point of this code change is changing the behavior of `to-stream`/`to-list` so that we can observe the internal structure of AVL-tree, right?

Reply:
I feel reluctant to give `to-stream`/`to-list` two concerns: one is converting data and another is observing the internal structure.

Although I'm not familiar with ML-style development, here are my ideas

* export testing-purpose functions in the module, where you declare that it's for testing purpose. Maybe we can have a convention like [stop special comment](https://ocaml.jp/refman/ch15s02.html#d4e12496) of ocamldoc
```
module TreeSet : sig
  %% Immutable set.
  type 'a t
  
   ...
  
  %%/%% 

  %% For testing
  %% ...
  val testing-to-list : 'a t -> 'a list :  'k 'v t -> ('k * 'v) list

end = struct ....
```

* Separate modules into `TreeSetImpl` and `TreeSet`, where `TreeSetImpl` exports low-level functions so that you can test them while `TreeSet` proxies high-level functionality of `TreeSetImpl`

The latter sounds more elegant to me while effort for code changes may not worth it. The former shouldn't require much code change and should be okay. Does it sound valid to you?

Reply:
> The point of this code change is changing the behavior of to-stream/to-list so that we can observe the internal structure of AVL-tree, right?

Right, that's what I wanted.

> I feel reluctant to give to-stream/to-list two concerns.

My intention was like:

1. The order of the results of `to-stream` and `to-list` is specified _unspecified_. By calling `to-stream` users can iterate over a set but they must not rely on its order. (In fact, it's ascending order.)
2. OTOH `generic.saty` knows the internal structure and implementation details of all modules including TreeSet. It uses the a priori knowledge that `to-stream` in fact yields elements in ascending order to test if the implementation is correct.

I think point 2 maybe sounds unnatural, because it makes the test suite more like a part of the implementation (white-box testing), instead of a neutral arbitrator standing outside the library (black-box testing). This unnaturality is somehow 仕方ない because of the limitation of the current library system (accessibility control system) of SATySFi.

So, we have two options here, one is to accept that `generic.saty` is white-box testing and change nothing, the other is to update the PR to rule out codes that uses domain-specific knowledge. I am neutral about both. If we choose the latter option I prefer the TreeSetImpl approach. What do you think?

Reply:
> In fact, it's ascending order
Oh, I thought the BFS order is different from an ascending order. Is it?

Reply:
hmm, I'm okay with leaving the current test. It's not ideal while I'm sure it shouldn't be a big issue. Maybe we can leave a comment in the test to tell that they are dependent on the unspecified behavior of `to-list`.

Reply:
> Oh, I thought the BFS order is different from an ascending order. Is it?

Sorry, this was incorrect. It *used to be* ascending order. Currently the order is that of the structure of the tree.

Reply:
> Maybe we can leave a comment in the test to tell that they are dependent on the unspecified behavior of to-list.

Sounds neat. I'm adding such a comment before merging.